File: androidApp/src/main/kotlin/com/aftcalculator/android/pdf/Form705Generator.kt
package com.aftcalculator.android.pdf

import android.content.Context
import com.aftcalculator.AftCalculator
import com.aftcalculator.models.AftEvent
import com.aftcalculator.models.AftScore
import com.aftcalculator.models.Gender
import com.aftcalculator.models.MosCategory
import com.tom_roush.pdfbox.android.PDFBoxResourceLoader
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDAcroForm
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDCheckBox
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDField
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDNonTerminalField
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDTerminalField
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlinx.coroutines.withTimeout
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

data class Form705Data(
    val soldierName: String,
    val graderName: String,
    val unitLocation: String,
    val mos: String,
    val payGrade: String,
    val score: AftScore,
    val testDate: Date = Date()
)

object Form705Generator {

    private var initialized = false

    fun initialize(context: Context) {
        if (!initialized) {
            PDFBoxResourceLoader.init(context)
            initialized = true
        }
    }

    suspend fun generateForm705(
        context: Context,
        data: Form705Data
    ): Result<File> = withContext(Dispatchers.IO) {
        initialize(context)

        try {
            android.util.Log.d("Form705", "Copying PDF from assets to cache...")

            // Copy the PDF to cache first - loading from file is more reliable
            // Always overwrite to ensure we have the latest template
            val templateFile = File(context.cacheDir, "da_form_705_template.pdf")
            context.assets.open("da_form_705.pdf").use { input ->
                templateFile.outputStream().use { output ->
                    input.copyTo(output)
                }
            }

            android.util.Log.d("Form705", "Loading PDF from file...")
            val document = PDDocument.load(templateFile)
            android.util.Log.d("Form705", "PDF loaded, pages: ${document.numberOfPages}")

            val acroForm = document.documentCatalog.acroForm

            if (acroForm != null) {
                android.util.Log.d("Form705", "Configuring AcroForm...")

                // Build a map of all fields by their fully qualified name
                val fieldMap = mutableMapOf<String, PDField>()
                val visited = mutableSetOf<String>()
                collectAllFields(acroForm.fields, fieldMap, visited, 0)

                android.util.Log.d("Form705", "Found ${fieldMap.size} total fields")

                android.util.Log.d("Form705", "Starting form fill...")
                val filledFields = fillForm(fieldMap, data)
                android.util.Log.d("Form705", "Form fill complete, filled ${filledFields.size} fields")

                // Generate appearances only for fields we filled (fast)
                // This keeps the rest of the form editable for graders
                android.util.Log.d("Form705", "Generating appearances for filled fields...")
                filledFields.forEach { field ->
                    try {
                        if (field is PDTerminalField) {
                            field.widgets.forEach { widget ->
                                // Force appearance generation for this widget
                                widget.getCOSObject().setNeedToBeUpdated(true)
                            }
                        }
                    } catch (e: Exception) {
                        android.util.Log.w("Form705", "Could not update appearance for ${field.fullyQualifiedName}")
                    }
                }
                android.util.Log.d("Form705", "Appearances generated")
            } else {
                android.util.Log.w("Form705", "No AcroForm found")
            }

            // Save to cache directory with soldier name and date
            val dateFormat = SimpleDateFormat("yyyyMMdd", Locale.US)
            val fileDate = dateFormat.format(data.testDate)

            // Parse soldier name (expected format: "Last, First, MI" or "Last, First")
            val nameParts = data.soldierName.split(",").map { it.trim() }
            val lastName = nameParts.getOrNull(0)?.uppercase()?.replace(" ", "_") ?: "UNKNOWN"
            val firstInitial = nameParts.getOrNull(1)?.firstOrNull()?.uppercaseChar() ?: 'X'

            val outputFile = File(context.cacheDir, "DA_Form_705_${lastName}_${firstInitial}_${fileDate}.pdf")
            android.util.Log.d("Form705", "Saving PDF to: ${outputFile.absolutePath}")

            // Use OutputStream for more control
            outputFile.outputStream().use { out ->
                document.save(out)
            }
            android.util.Log.d("Form705", "PDF saved, closing document...")

            document.close()

            android.util.Log.d("Form705", "PDF complete, size: ${outputFile.length()} bytes")

            Result.success(outputFile)
        } catch (e: Exception) {
            android.util.Log.e("Form705", "Error generating PDF", e)
            Result.failure(e)
        }
    }

    private fun collectAllFields(
        fields: List<PDField>?,
        fieldMap: MutableMap<String, PDField>,
        visited: MutableSet<String> = mutableSetOf(),
        depth: Int = 0
    ) {
        // Prevent infinite recursion
        if (depth > 10 || fields == null) return

        fields.forEach { field ->
            val name = field.fullyQualifiedName

            // Skip if already visited (prevents circular references)
            if (name in visited) return@forEach
            visited.add(name)

            fieldMap[name] = field

            // If it's a non-terminal field, recurse into children (with depth limit)
            if (field is PDNonTerminalField && depth < 10) {
                collectAllFields(field.children, fieldMap, visited, depth + 1)
            }
        }
    }

    private fun fillForm(fieldMap: Map<String, PDField>, data: Form705Data): List<PDField> {
        val filledFields = mutableListOf<PDField>()
        val dateFormat = SimpleDateFormat("yyyyMMdd", Locale.US)
        val formattedDate = dateFormat.format(data.testDate)

        android.util.Log.d("Form705", "Filling header fields...")

        // Header fields (Page 1)
        setField(fieldMap, "form1[0].Page1[0].Name[0]", data.soldierName)?.let { filledFields.add(it) }
        setField(fieldMap, "form1[0].Page1[0].Unit_Location[0]", data.unitLocation)?.let { filledFields.add(it) }

        // Check Combat or General standard based on MOS category
        android.util.Log.d("Form705", "Setting standard checkbox for ${data.score.soldier.mosCategory}...")
        when (data.score.soldier.mosCategory) {
            MosCategory.COMBAT -> {
                setCheckbox(fieldMap, "form1[0].Page1[0].Check_Standard_Combat[0]", true)?.let { filledFields.add(it) }
            }
            MosCategory.COMBAT_ENABLING -> {
                setCheckbox(fieldMap, "form1[0].Page1[0].Check_Standard_General[0]", true)?.let { filledFields.add(it) }
            }
        }

        // Set gender checkbox
        android.util.Log.d("Form705", "Setting gender checkbox for ${data.score.soldier.gender}...")
        when (data.score.soldier.gender) {
            Gender.MALE -> {
                setCheckbox(fieldMap, "form1[0].Page1[0].Male[0]", true)?.let { filledFields.add(it) }
            }
            Gender.FEMALE -> {
                setCheckbox(fieldMap, "form1[0].Page1[0].Female[0]", true)?.let { filledFields.add(it) }
            }
        }

        android.util.Log.d("Form705", "Filling Test One section...")

        // Test One section
        setField(fieldMap, "form1[0].Page1[0].Test_One_Date[0]", formattedDate)?.let { filledFields.add(it) }
        setField(fieldMap, "form1[0].Page1[0].Test_One_MOS[0]", data.mos)?.let { filledFields.add(it) }
        setField(fieldMap, "form1[0].Page1[0].Test_One_Rank_Grade[0]", data.payGrade)?.let { filledFields.add(it) }
        setField(fieldMap, "form1[0].Page1[0].Test_One_Age[0]", data.score.soldier.age.toString())?.let { filledFields.add(it) }

        android.util.Log.d("Form705", "Filling event scores...")

        // Fill event scores (text fields only - skip checkboxes for now)
        data.score.eventScores.forEach { eventScore ->
            filledFields.addAll(fillEventScore(fieldMap, eventScore.event, eventScore.rawValue, eventScore.points))
        }

        android.util.Log.d("Form705", "Filling totals...")

        // Total points
        setField(fieldMap, "form1[0].Page1[0].Test_One_Total_Points[0]", data.score.totalPoints.toString())?.let { filledFields.add(it) }

        // Grader info
        setField(fieldMap, "form1[0].Page1[0].OIC_NCOIC_Name_Test_One[0]", data.graderName)?.let { filledFields.add(it) }
        setField(fieldMap, "form1[0].Page1[0].OIC_NCOIC_Date_Test_One[0]", formattedDate)?.let { filledFields.add(it) }

        android.util.Log.d("Form705", "Form fill function complete")
        return filledFields
    }

    private fun fillEventScore(fieldMap: Map<String, PDField>, event: AftEvent, rawValue: Double, points: Int): List<PDField> {
        val filledFields = mutableListOf<PDField>()
        val prefix = "form1[0].Page1[0]."

        when (event) {
            AftEvent.DEADLIFT -> {
                setField(fieldMap, "${prefix}Test_One_First_Attempt[0]", rawValue.toInt().toString())?.let { filledFields.add(it) }
                // Skip checkbox: setCheckbox(fieldMap, "${prefix}Test_One_First_Attempt_Check[0]", true)
                setField(fieldMap, "${prefix}Test_One_Points1[0]", points.toString())?.let { filledFields.add(it) }
            }
            AftEvent.PUSH_UP -> {
                setField(fieldMap, "${prefix}Test_One_Repetitions[0]", rawValue.toInt().toString())?.let { filledFields.add(it) }
                setField(fieldMap, "${prefix}Test_One_Points3[0]", points.toString())?.let { filledFields.add(it) }
            }
            AftEvent.SPRINT_DRAG_CARRY -> {
                setField(fieldMap, "${prefix}Test_One_Time1[0]", AftCalculator.formatTime(rawValue))?.let { filledFields.add(it) }
                setField(fieldMap, "${prefix}Test_One_Points4[0]", points.toString())?.let { filledFields.add(it) }
            }
            AftEvent.PLANK -> {
                setField(fieldMap, "${prefix}Test_One_Time2[0]", AftCalculator.formatTime(rawValue))?.let { filledFields.add(it) }
                setField(fieldMap, "${prefix}Test_One_Points5[0]", points.toString())?.let { filledFields.add(it) }
            }
            AftEvent.TWO_MILE_RUN -> {
                setField(fieldMap, "${prefix}Test_One_Time3[0]", AftCalculator.formatTime(rawValue))?.let { filledFields.add(it) }
                setField(fieldMap, "${prefix}Test_One_Points6[0]", points.toString())?.let { filledFields.add(it) }
            }
            // Alternate aerobic events use Row_Swim_Bike_Walk field
            AftEvent.WALK_2_5_MILE, AftEvent.ROW_5K, AftEvent.BIKE_12K, AftEvent.SWIM_1K -> {
                // Fill the alternate event description and time
                val eventDesc = when (event) {
                    AftEvent.WALK_2_5_MILE -> "2.5 MI WALK"
                    AftEvent.ROW_5K -> "5K ROW"
                    AftEvent.BIKE_12K -> "12K BIKE"
                    AftEvent.SWIM_1K -> "1K SWIM"
                    else -> event.displayName
                }
                setField(fieldMap, "${prefix}Test_One_Row_Swim_Bike_Walk[0]", eventDesc)?.let { filledFields.add(it) }
                if (rawValue > 0) {
                    setField(fieldMap, "${prefix}Test_One_Time4[0]", AftCalculator.formatTime(rawValue))?.let { filledFields.add(it) }
                }
                setField(fieldMap, "${prefix}Test_One_Points7[0]", points.toString())?.let { filledFields.add(it) }
            }
        }
        return filledFields
    }

    private fun setField(fieldMap: Map<String, PDField>, fieldName: String, value: String): PDField? {
        val field = fieldMap[fieldName]
        if (field != null) {
            try {
                field.setValue(value)
                android.util.Log.d("Form705", "✓ Set '$fieldName' = '$value'")
                return field
            } catch (e: Exception) {
                android.util.Log.e("Form705", "✗ Error setting '$fieldName': ${e.message}")
            }
        } else {
            android.util.Log.w("Form705", "✗ Field not found: $fieldName")
        }
        return null
    }

    private fun setCheckbox(fieldMap: Map<String, PDField>, fieldName: String, checked: Boolean): PDField? {
        val field = fieldMap[fieldName]
        if (field != null) {
            try {
                if (field is PDCheckBox) {
                    if (checked) field.check() else field.unCheck()
                    android.util.Log.d("Form705", "✓ Checkbox '$fieldName' = $checked")
                } else {
                    // For radio buttons or other button types, try setting value
                    val onValue = getOnValue(field)
                    field.setValue(if (checked) onValue else "Off")
                    android.util.Log.d("Form705", "✓ Button '$fieldName' = ${if (checked) onValue else "Off"}")
                }
                return field
            } catch (e: Exception) {
                android.util.Log.e("Form705", "✗ Error setting checkbox '$fieldName': ${e.message}")
            }
        } else {
            android.util.Log.w("Form705", "✗ Checkbox not found: $fieldName")
        }
        return null
    }

    private fun getOnValue(field: PDField): String {
        // Try to get the "on" value for a checkbox/radio button
        return try {
            if (field is PDTerminalField) {
                val widgets = field.widgets
                if (widgets.isNotEmpty()) {
                    val appearance = widgets[0].appearance
                    val normalAppearance = appearance?.normalAppearance
                    if (normalAppearance != null && !normalAppearance.isSubDictionary) {
                        // Get available appearance states
                        val dict = normalAppearance
                        // Return first non-Off value
                        return "Yes" // Default fallback
                    }
                }
            }
            "Yes"
        } catch (e: Exception) {
            "Yes"
        }
    }

    fun getFormFieldNames(context: Context): List<String> {
        initialize(context)

        return try {
            val inputStream = context.assets.open("da_form_705.pdf")
            val document = PDDocument.load(inputStream)
            val acroForm = document.documentCatalog.acroForm

            val fieldMap = mutableMapOf<String, PDField>()
            val visited = mutableSetOf<String>()
            collectAllFields(acroForm?.fields, fieldMap, visited, 0)

            document.close()
            inputStream.close()

            fieldMap.keys.sorted()
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }
}

File: iosApp/iosApp/Form705Generator.swift
import Foundation
import PDFKit
import UIKit

struct Form705Data {
    let soldierName: String
    let graderName: String
    let unitLocation: String
    let mos: String
    let payGrade: String
    let score: SwiftAftScore
    let testDate: Date

    init(
        soldierName: String,
        graderName: String,
        unitLocation: String,
        mos: String,
        payGrade: String,
        score: SwiftAftScore,
        testDate: Date = Date()
    ) {
        self.soldierName = soldierName
        self.graderName = graderName
        self.unitLocation = unitLocation
        self.mos = mos
        self.payGrade = payGrade
        self.score = score
        self.testDate = testDate
    }
}

class Form705Generator {

    // Cache for preloaded template
    private static var cachedTemplateURL: URL?
    private static var cachedDocument: PDFDocument?

    /// Preload the PDF template to avoid first-use delays
    static func preloadTemplate() {
        DispatchQueue.global(qos: .userInitiated).async {
            if cachedTemplateURL == nil {
                cachedTemplateURL = Bundle.main.url(forResource: "da_form_705", withExtension: "pdf")
                if let url = cachedTemplateURL {
                    // Load the document once to warm up PDFKit
                    let _ = PDFDocument(url: url)
                    print("Form705: Template preloaded")
                }
            }
        }
    }

    static func generateForm705(data: Form705Data) -> Result<URL, Error> {
        // Use cached URL if available, otherwise search bundle
        var templateURL: URL? = cachedTemplateURL

        if templateURL == nil {
            // Debug: List all PDF files in bundle
            print("Form705: Searching for PDF in bundle...")
            if let bundlePath = Bundle.main.resourcePath {
                print("Form705: Bundle path: \(bundlePath)")
                let fileManager = FileManager.default
                if let contents = try? fileManager.contentsOfDirectory(atPath: bundlePath) {
                    let pdfFiles = contents.filter { $0.hasSuffix(".pdf") }
                    print("Form705: PDF files in bundle: \(pdfFiles)")
                    if pdfFiles.isEmpty {
                        print("Form705: No PDF files found in bundle!")
                        // List all files for debugging
                        print("Form705: All bundle files: \(contents.prefix(30))...")
                    }
                }
                // Also check for Resources subfolder
                let resourcesPath = bundlePath + "/Resources"
                if let resourceContents = try? fileManager.contentsOfDirectory(atPath: resourcesPath) {
                    let resourcePdfs = resourceContents.filter { $0.hasSuffix(".pdf") }
                    print("Form705: PDF files in Resources subfolder: \(resourcePdfs)")
                }
            }

            // Get the template PDF from bundle - try multiple locations
            templateURL = Bundle.main.url(forResource: "da_form_705", withExtension: "pdf")

            // If not found, try with subdirectory
            if templateURL == nil {
                templateURL = Bundle.main.url(forResource: "da_form_705", withExtension: "pdf", subdirectory: "Resources")
                if templateURL != nil {
                    print("Form705: Found template in Resources subdirectory")
                }
            }

            // Cache for next time
            cachedTemplateURL = templateURL
        } else {
            print("Form705: Using cached template URL")
        }

        guard let templateURL = templateURL else {
            print("Form705: Template not found in bundle - PDF must be added to Xcode target")
            return .failure(Form705Error.templateNotFound)
        }

        guard let document = PDFDocument(url: templateURL) else {
            print("Form705: Cannot load PDF document")
            return .failure(Form705Error.cannotLoadTemplate)
        }

        guard let page = document.page(at: 0) else {
            print("Form705: Cannot access page 0")
            return .failure(Form705Error.cannotAccessPage)
        }

        // Get annotations (form fields)
        let annotations = page.annotations
        print("Form705: Found \(annotations.count) annotations")

        // Date formatter for the form
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyyMMdd"
        let formattedDate = dateFormatter.string(from: data.testDate)

        var filledFields = 0

        // Fill form fields
        for annotation in annotations {
            guard let fieldName = annotation.fieldName else { continue }

            // Header fields
            if fieldName.contains("Name[0]") && !fieldName.contains("OIC") {
                annotation.widgetStringValue = data.soldierName
                filledFields += 1
            }
            else if fieldName.contains("Unit_Location[0]") {
                annotation.widgetStringValue = data.unitLocation
                filledFields += 1
            }
            // Standard checkboxes
            else if fieldName.contains("Check_Standard_Combat[0]") {
                if data.score.isCombatMos {
                    annotation.buttonWidgetState = .onState
                    filledFields += 1
                }
            }
            else if fieldName.contains("Check_Standard_General[0]") {
                if !data.score.isCombatMos {
                    annotation.buttonWidgetState = .onState
                    filledFields += 1
                }
            }
            // Gender checkboxes
            else if fieldName.contains("Male[0]") {
                if data.score.soldierGender == "Male" {
                    annotation.buttonWidgetState = .onState
                    filledFields += 1
                }
            }
            else if fieldName.contains("Female[0]") {
                if data.score.soldierGender == "Female" {
                    annotation.buttonWidgetState = .onState
                    filledFields += 1
                }
            }
            // Test One fields
            else if fieldName.contains("Test_One_Date[0]") {
                annotation.widgetStringValue = formattedDate
                filledFields += 1
            }
            else if fieldName.contains("Test_One_MOS[0]") {
                annotation.widgetStringValue = data.mos
                filledFields += 1
            }
            else if fieldName.contains("Test_One_Rank_Grade[0]") {
                annotation.widgetStringValue = data.payGrade
                filledFields += 1
            }
            else if fieldName.contains("Test_One_Age[0]") {
                annotation.widgetStringValue = "\(data.score.soldierAge)"
                filledFields += 1
            }
            // Event scores
            else if fieldName.contains("Test_One_First_Attempt[0]") {
                if let deadlift = data.score.eventScores.first(where: { $0.eventName.contains("Deadlift") }) {
                    annotation.widgetStringValue = "\(Int(deadlift.rawValue))"
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Points1[0]") {
                if let deadlift = data.score.eventScores.first(where: { $0.eventName.contains("Deadlift") }) {
                    annotation.widgetStringValue = "\(deadlift.points)"
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Repetitions[0]") {
                if let pushUp = data.score.eventScores.first(where: { $0.eventName.contains("Push") }) {
                    annotation.widgetStringValue = "\(Int(pushUp.rawValue))"
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Points3[0]") {
                if let pushUp = data.score.eventScores.first(where: { $0.eventName.contains("Push") }) {
                    annotation.widgetStringValue = "\(pushUp.points)"
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Time1[0]") {
                if let sdc = data.score.eventScores.first(where: { $0.eventName.contains("Sprint") }) {
                    annotation.widgetStringValue = formatTime(seconds: sdc.rawValue)
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Points4[0]") {
                if let sdc = data.score.eventScores.first(where: { $0.eventName.contains("Sprint") }) {
                    annotation.widgetStringValue = "\(sdc.points)"
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Time2[0]") {
                if let plank = data.score.eventScores.first(where: { $0.eventName.contains("Plank") }) {
                    annotation.widgetStringValue = formatTime(seconds: plank.rawValue)
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Points5[0]") {
                if let plank = data.score.eventScores.first(where: { $0.eventName.contains("Plank") }) {
                    annotation.widgetStringValue = "\(plank.points)"
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Time3[0]") {
                if let run = data.score.eventScores.first(where: { $0.eventName.contains("2 Mile") }) {
                    annotation.widgetStringValue = formatTime(seconds: run.rawValue)
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Points6[0]") {
                if let run = data.score.eventScores.first(where: { $0.eventName.contains("2 Mile") }) {
                    annotation.widgetStringValue = "\(run.points)"
                    filledFields += 1
                }
            }
            // Alternate aerobic event fields
            else if fieldName.contains("Test_One_Row_Swim_Bike_Walk[0]") {
                if let alternate = findAlternateAerobicEvent(in: data.score.eventScores) {
                    annotation.widgetStringValue = alternate.eventName
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Time4[0]") {
                if let alternate = findAlternateAerobicEvent(in: data.score.eventScores) {
                    annotation.widgetStringValue = formatTime(seconds: alternate.rawValue)
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Points7[0]") {
                if let alternate = findAlternateAerobicEvent(in: data.score.eventScores) {
                    annotation.widgetStringValue = "\(alternate.points)"
                    filledFields += 1
                }
            }
            // Totals
            else if fieldName.contains("Test_One_Total_Points[0]") {
                annotation.widgetStringValue = "\(data.score.totalPoints)"
                filledFields += 1
            }
            // GO / NO GO checkboxes for alternate aerobic event row
            // Only check if an alternate event was used
            else if fieldName.contains("Test_One_Go[0]") && !fieldName.contains("NoGo") && !fieldName.contains("Final") && !fieldName.contains("Weight") && !fieldName.contains("Body_Fat") {
                if let altEvent = findAlternateAerobicEvent(in: data.score.eventScores) {
                    if altEvent.points >= 60 {
                        annotation.buttonWidgetState = .onState
                        filledFields += 1
                    }
                }
            }
            else if fieldName.contains("Test_One_NoGo[0]") && !fieldName.contains("Final") && !fieldName.contains("Weight") && !fieldName.contains("Body_Fat") {
                if let altEvent = findAlternateAerobicEvent(in: data.score.eventScores) {
                    if altEvent.points < 60 {
                        annotation.buttonWidgetState = .onState
                        filledFields += 1
                    }
                }
            }
            // Final GO / NO GO checkboxes for overall test result
            else if fieldName.contains("Test_One_Final_Go[0]") {
                if data.score.passed {
                    annotation.buttonWidgetState = .onState
                    filledFields += 1
                }
            }
            else if fieldName.contains("Test_One_Final_NoGo[0]") {
                if !data.score.passed {
                    annotation.buttonWidgetState = .onState
                    filledFields += 1
                }
            }
            // Grader info
            else if fieldName.contains("OIC_NCOIC_Name_Test_One[0]") {
                annotation.widgetStringValue = data.graderName
                filledFields += 1
            }
            else if fieldName.contains("OIC_NCOIC_Date_Test_One[0]") {
                annotation.widgetStringValue = formattedDate
                filledFields += 1
            }
        }

        print("Form705: Filled \(filledFields) fields")

        // Generate output filename
        let outputURL = getOutputURL(data: data)

        // Remove existing file if present
        try? FileManager.default.removeItem(at: outputURL)

        // Use document.write(to:) to preserve form field editability
        // The "structure tree" warning can be ignored - it doesn't prevent saving
        if document.write(to: outputURL) {
            print("Form705: PDF written using document.write to \(outputURL.path)")

            // Small delay to ensure file system sync
            Thread.sleep(forTimeInterval: 0.1)

            // Verify file was created and has content
            if FileManager.default.fileExists(atPath: outputURL.path) {
                let fileSize = (try? FileManager.default.attributesOfItem(atPath: outputURL.path)[.size] as? Int) ?? 0
                print("Form705: File size: \(fileSize) bytes")
                if fileSize > 1000 {  // Ensure file has meaningful content
                    return .success(outputURL)
                } else {
                    print("Form705: File too small, might be corrupted")
                }
            }
        }

        // Fallback: try dataRepresentation
        print("Form705: document.write failed, trying dataRepresentation...")
        if let pdfData = document.dataRepresentation() {
            do {
                try pdfData.write(to: outputURL)
                print("Form705: PDF written using dataRepresentation to \(outputURL.path)")

                if FileManager.default.fileExists(atPath: outputURL.path) {
                    let fileSize = (try? FileManager.default.attributesOfItem(atPath: outputURL.path)[.size] as? Int) ?? 0
                    print("Form705: File size: \(fileSize) bytes")
                    return .success(outputURL)
                }
            } catch {
                print("Form705: dataRepresentation write failed: \(error)")
            }
        }

        print("Form705: All save methods failed")
        return .failure(Form705Error.cannotSavePDF)
    }

    private static func getOutputURL(data: Form705Data) -> URL {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyyMMdd"
        let fileDate = dateFormatter.string(from: data.testDate)

        let nameParts = data.soldierName.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
        let lastName = nameParts.first?.uppercased().replacingOccurrences(of: " ", with: "_") ?? "UNKNOWN"
        let firstInitial = nameParts.count > 1 ? String(nameParts[1].prefix(1)).uppercased() : "X"

        let fileName = "DA_Form_705_\(lastName)_\(firstInitial)_\(fileDate).pdf"

        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        return documentsURL.appendingPathComponent(fileName)
    }

    private static func formatTime(seconds: Double) -> String {
        let totalSeconds = Int(seconds)
        let minutes = totalSeconds / 60
        let secs = totalSeconds % 60
        return "\(minutes):\(String(format: "%02d", secs))"
    }

    private static func findAlternateAerobicEvent(in eventScores: [SwiftEventScore]) -> SwiftEventScore? {
        return eventScores.first { score in
            score.eventName.contains("Walk") ||
            score.eventName.contains("Row") ||
            score.eventName.contains("Bike") ||
            score.eventName.contains("Swim")
        }
    }
}

enum Form705Error: LocalizedError {
    case templateNotFound
    case cannotLoadTemplate
    case cannotAccessPage
    case cannotSavePDF

    var errorDescription: String? {
        switch self {
        case .templateNotFound:
            return "DA Form 705 template not found. In Xcode: 1) Select da_form_705.pdf in Project Navigator, 2) In File Inspector, check 'Target Membership: iosApp', 3) Clean build (Cmd+Shift+K) and rebuild."
        case .cannotLoadTemplate:
            return "Cannot load DA Form 705 template - the PDF file may be corrupted"
        case .cannotAccessPage:
            return "Cannot access PDF page - template may be invalid"
        case .cannotSavePDF:
            return "Cannot save filled PDF - check device storage"
        }
    }
}
